main.py

```diff
@@ -20,208 +22,295 @@
+    semaphore: asyncio.Semaphore,
+    max_retries: int,
+    ytdlp_downloader: callable,
+    loop: asyncio.AbstractEventLoop,
+) -> Any:
+    """
+    Executes a work function asynchronously with a semaphore.
 
-    if COOKIE_FILE != None:
-        opts.pop("cookiesfrombrowser")
-    else:
-        opts.pop("cookiefile")
+    Args:
+        url (str): The URL to be processed.
+        semaphore (asyncio.Semaphore): The semaphore to limit the number of concurrent workers.
+        max_retries (int): The maximum number of retries for the work function.
+        ytdlp_downloader (callable): a callable ref to the yt_dlp.YoutubeDL.download() method.
+        loop (asyncio.AbstractEventLoop): The event loop to run the work function.
 
-    with yt_dlp.YoutubeDL(opts) as ydl:
-        for i, event in enumerate(events):
-            timestamp = event["time"]
-            url = event["titleUrl"]
-            title = event["title"][8:]
-
-            marker_path = os.path.join(
-                DONE_DIRECTORY, hashlib.sha256(url.encode("utf-8")).hexdigest()
-            )
-
-            print(
-                f"{i}/{len(events)} \t {timestamp} \t {url} \t {title} ... ",
-                end="",
-                flush=True,
-            )
-
-            if os.path.exists(marker_path):
-                print(" -> Already done")
-                continue
-
-            try:
-                ydl.download(url)
-                print(" -> Sleeping ... ", end="", flush=True)
-                time.sleep(SLEEP_MIN + random.random()
-                           * (SLEEP_MAX - SLEEP_MIN))
-                print(" -> Done")
-            except yt_dlp.utils.DownloadError:
-                print(" -> DownloadError")
-
-            with open(marker_path, "w"):
-                pass
+    Returns:
+        Any: The result of the work function.
+
+    """
+    async with semaphore:
+        async with aiofiles.open("history.log", "a+") as f:
+
+            await f.seek(0)
+            processed: list[str] = [line.rstrip() for line in await f.readlines()]
+
+            logger.info(f"Processing URL: {url}")
+
+            if url in processed:
+                logger.info(f"URL {url} has already been processed. Skipping...")
+                return
+
+            for attempt in range(max_retries):
+                try:
+                    result: Any = await loop.run_in_executor(
+                        None, lambda: ytdlp_downloader(url)
+                    )
+                    await f.write(url + "\n")
+                    break
+                except Exception as e:
+                    logger.warning(f"Attempt {attempt + 1} failed: {e}. Retrying...")
+                    if attempt == max_retries - 1:
+                        logger.error(f"Failed after {max_retries} attempts.")
+                        result = None
+                        break
+
+            await asyncio.sleep(random.uniform(1, 3))
+            logger.info(f"Processed URL: {url}.")
```

Comment:
Add counter

Reply:
If you can, contribute that change to the PR. Thank you!

Reply:
I don't know how to do it. Maybe a task_done callback function can do. Will it also count the errors?